### ISSUE TITLE
Profoundly unhelpful compiler error message when "mod" accidentally used as "%"

### ISSUE THREAD

**Issue by mcolinj** [rust-lang/rust#45560]
```
error: Could not compile `leap`.
warning: build failed, waiting for other jobs to finish...
error[E0308]: mismatched types
 --> src/lib.rs:1:40
  |
1 |   pub fn is_leap_year(year: i32) -> bool {
  |  ________________________________________^
2 | |     if year mod 4 == 0 {
3 | |         true
4 | |     } else {
5 | |         false
6 | |     }
7 | | }
  | |_^ expected bool, found ()
  |
  = note: expected type `bool`
             found type `()`

error: aborting due to 2 previous errors

error: Could not compile `leap`.

```

**Comment by scottmcm** [comment]
Note that [it](https://play.rust-lang.org/?gist=8a3b141d1c95f4e41105dfa3cea02bae&version=stable) also gives this as the first error:
```rust
error: expected `{`, found `mod`
 --> src/main.rs:2:13
  |
2 |     if year mod 4 == 0 {
  |             ^^^
```

**Comment by mcolinj** [comment]
Yes, thanks.   I see that now that I've reproduced it.    That first error scrolled off the screen so I did not see it!    I'm a newbie at Rust Belt Rust today and my workshop instructor suggest I file this PR after see it.

**Comment by scottmcm** [comment]
Thanks for filing!  There's definitely opportunity for improvement here -- perhaps type errors should never be shown for methods that failed to parse, or something.

**Comment by estebank** [comment]
Fixed in #46732. [The compiler now only complains about `mod`](https://play.rust-lang.org/?gist=8b8d1b3dac4532ee789fc83d79f7b642&version=beta) and won't try to type check blocks that have had parse errors:

```
error: expected `{`, found `mod`
 --> src/main.rs:2:13
  |
2 |     if year mod 4 == 0 {
  |             ^^^

error: aborting due to previous error
```
